// react-native-thermal-pos-printer/lib/commonjs/utils/debug.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PosPrinterDebug = void 0;
exports.testNativeModule = testNativeModule;
var _native = require("./native.js");
var _reactNative = require("react-native");
class PosPrinterDebug {
  static log(message, ...args) {
    if (__DEV__) {
      console.log(`[PosPrinter] ${message}`, ...args);
    }
  }
  static error(message, ...args) {
    if (__DEV__) {
      console.error(`[PosPrinter] ${message}`, ...args);
    }
  }
  static checkModuleRegistration() {
    const moduleAvailable = !!_reactNative.NativeModules.PosPrinter;
    const availableMethods = moduleAvailable ? Object.keys(_reactNative.NativeModules.PosPrinter).filter(key => typeof _reactNative.NativeModules.PosPrinter[key] === 'function') : [];
    const debugInfo = {
      moduleAvailable,
      platform: _reactNative.Platform.OS,
      availableMethods,
      nativeModules: Object.keys(_reactNative.NativeModules)
    };
    console.log('Module Registration Debug:', debugInfo);
    if (!debugInfo.moduleAvailable) {
      console.error('❌ Native module not found!');
      console.log('Available native modules:', debugInfo.nativeModules);
      console.log('Platform:', debugInfo.platform);
    } else {
      console.log('✅ Native module is available');
      console.log('Available methods:', debugInfo.availableMethods);
    }
  }
}
exports.PosPrinterDebug = PosPrinterDebug;
async function testNativeModule() {
  try {
    console.log('Testing native module availability...');
    const result = await _native.PosPrinter.init({});
    console.log('Init result:', result);
    return true;
  } catch (error) {
    console.error('Native module test failed:', error);
    return false;
  }
}
